### ISSUE TITLE
Parsing quoted strings in response body

### ISSUE THREAD

**Issue by tobiashm** [apiaryio/redsnow#63]
Given I specify a JSON response with a string containing quotation marks, or other escaped characters,
then `RedSnow.parse` fails with a `JSON::ParseError`

Example:
```markdown
FORMAT: 1A

# GET /escaped

+ Response 200 (application/json)

        {
            "text": "foo \"bar\" baz\n"
        }
```

This apparently generates a JSON document with, among other, the following content: 
`"body": "{\n    \"text\": \"foo \\"bar\\" baz\n\"\n}\n"` which will not parse.

Stack trace:
```
	from .../gems/json-1.8.2/lib/json/common.rb:155:in `parse'
	from .../gems/json-1.8.2/lib/json/common.rb:155:in `parse'
	from .../gems/redsnow-0.4.1/lib/redsnow/parseresult.rb:24:in `initialize'
	from .../gems/redsnow-0.4.1/lib/redsnow.rb:47:in `new'
	from .../gems/redsnow-0.4.1/lib/redsnow.rb:47:in `parse'
```

**Comment by tobiashm** [comment]
As far as I can tell, this problem was introduced in 0.4.0, and seems to be related to changes in  snowcrash 1.0.0

I have not been able to get a failure in other tools; e.g. drafter(.js)
But it does seem like it stems from the escaping of the strings coming out of snowcrash.

**Comment by zdne** [comment]
Thanks @tobiashm! We will look into this.

**Comment by tobiashm** [comment]
This is the output from Drafter parsing the blueprint (the `RedSnow::Binding.drafter_c_parse` call). As it can be seen, it's not valid JSON:
```json
{
  "_version": "2.1",
  "ast": {
    "_version": "3.0",
    "metadata": [
      {
        "name": "FORMAT",
        "value": "1A"
      }
    ],
    "name": "",
    "description": "",
    "element": "category",
    "resourceGroups": [
      {
        "name": "",
        "description": "",
        "resources": [
          {
            "element": "resource",
            "name": "",
            "description": "",
            "uriTemplate": "/escaped",
            "model": {},
            "parameters": [],
            "actions": [
              {
                "name": "",
                "description": "",
                "method": "GET",
                "parameters": [],
                "attributes": {
                  "relation": "",
                  "uriTemplate": ""
                },
                "content": [],
                "examples": [
                  {
                    "name": "",
                    "description": "",
                    "requests": [],
                    "responses": [
                      {
                        "name": "200",
                        "description": "",
                        "headers": [
                          {
                            "name": "Content-Type",
                            "value": "application/json"
                          }
                        ],
                        "body": "{\n    \"text\": \"foo \\"bar\\" baz\n\"\n}\n",
                        "schema": "",
                        "content": [
                          {
                            "element": "asset",
                            "attributes": {
                              "role": "bodyExample"
                            },
                            "content": "{\n    \"text\": \"foo \\"bar\\" baz\n\"\n}\n"
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ],
            "content": []
          }
        ]
      }
    ],
    "content": [
      {
        "element": "category",
        "content": [
          {
            "element": "resource",
            "name": "",
            "description": "",
            "uriTemplate": "/escaped",
            "model": {},
            "parameters": [],
            "actions": [
              {
                "name": "",
                "description": "",
                "method": "GET",
                "parameters": [],
                "attributes": {
                  "relation": "",
                  "uriTemplate": ""
                },
                "content": [],
                "examples": [
                  {
                    "name": "",
                    "description": "",
                    "requests": [],
                    "responses": [
                      {
                        "name": "200",
                        "description": "",
                        "headers": [
                          {
                            "name": "Content-Type",
                            "value": "application/json"
                          }
                        ],
                        "body": "{\n    \"text\": \"foo \\"bar\\" baz\n\"\n}\n",
                        "schema": "",
                        "content": [
                          {
                            "element": "asset",
                            "attributes": {
                              "role": "bodyExample"
                            },
                            "content": "{\n    \"text\": \"foo \\"bar\\" baz\n\"\n}\n"
                          }
                        ]
                      }
                    ]
                  }
                ]
              }
            ],
            "content": []
          }
        ]
      }
    ]
  },
  "error": {
    "code": 0,
    "message": "",
    "location": []
  }
}
```
(Using RedSnow version 0.4.1)

**Comment by tobiashm** [comment]
Getting similar result (same invalid JSON) using Drafter CLI version 0.1.8

**Comment by zdne** [comment]
> Getting similar result (same invalid JSON) using Drafter CLI version 0.1.8

@tobiashm so it seems the issue is with Drafter itself, correct?

**Comment by tobiashm** [comment]
Yes, that would be my conclusion. Haven't tried the snowcrash lib directly.

**Comment by tobiashm** [comment]
Now I've also tried out protagonist and drafter.js, and they both generate correctly formatted output. And they're also based on the Drafter lib.

Running drafter.js from the command-line, outputs JSON:
```json
                        "body": "{\n    \"text\": \"foo \\\"bar\\\" baz\\n\"\n}\n",
```

Using Protagonist in Node to parse:
```js
blueprint = require('fs').readFileSync('test.apib').toString();
result = require('protagonist').parseSync(blueprint);
result.ast.content[0].content[0].actions[0].examples[0].responses[0].body // =>
'{\n    "text": "foo \\"bar\\" baz\\n"\n}\n'
```

**Comment by zdne** [comment]
Must be a bug in the Drafter (C++) itself, if I am correct the protagonist / Drafter.js are not using C++ Drafter for serialization only for parsing.

**Comment by pksunkara** [comment]
Maybe it's a bug in escaping in the [sos](https://github.com/apiaryio/sos) library

**Comment by tobiashm** [comment]
Sounds plausible that it's a problem with the serialisation, as the object representation from the parser in Protaginist is correct, and drafter.js serialises (a modified version of that) with `JSON.stringify`.

**Comment by tobiashm** [comment]
@pksunkara looks like you're right about sos.

I've added a double-escaped string to the tests, which shows the problem: https://github.com/apiaryio/sos/pull/7

**Comment by pksunkara** [comment]
Thanks for the test. We need to fix that.

**Comment by zdne** [comment]
Thanks @tobiashm , this is tremendous help!

**Comment by tobiashm** [comment]
@zdne @pksunkara thank you for a great system. We're only getting started with API Blueprints, but it's been an enormous help so far.

**Comment by pksunkara** [comment]
Fixed in redsnow 0.4.2
